### PR TITLE
feat(eslint-plugin): [await-thenable] report unnecessary `await using` statements

### PR DESCRIPTION
**Pull request by kirkwaiblinger**
## PR Checklist

- [x] Addresses an existing open issue: fixes part of #10208 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Basically just checks whether the RHS of an `await using` statement _might_ have a `[Symbol.asyncDispose]()`, otherwise reports. 

Added in docs and a couple minor refactors. Also modified the `getFixOrSuggest()` method to allow providing _no_ fix or suggestion.

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/await-thenable.test.ts [hunk 1]
```diff
@@ -475,5 +518,99 @@ for  (const value of yieldNumberPromises()) {
...
+  };
+}
+      `,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
```
Comment (kirkwaiblinger) [inline packages/eslint-plugin/tests/rules/await-thenable.test.ts:580]: Honestly I'm kind of torn on this one. In principle, the user could split it up to be 
```ts
  using a = disposable,
  await using b = asyncDisposable,
  using c = disposable,
  await using d = asyncDisposable,
  using e = disposable;
```

But it doesn't really have any benefit (unlike how the `for await` case can be actively harmful to use `for await` unnecessarily). So we could change the logic to make this a valid case unless _none_ of the declarators are possibly-async-iterable 🤷

File: packages/eslint-plugin/src/rules/await-thenable.ts [hunk 2]
```diff
@@ -112,6 +116,57 @@ export default createRule<[], MessageId>({
           });
         }
       },
+
+      'VariableDeclaration[kind="await using"]'(
+        node: TSESTree.VariableDeclaration,
+      ): void {
+        for (const declarator of node.declarations) {
+          const init = nullThrows(
```
Comment (kirkwaiblinger) [inline packages/eslint-plugin/src/rules/await-thenable.ts:124]: Turns out this is not sound, see 
```ts
async function awaitUsingLoopInitializer() {
  for (await using x of []) {}
}
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 3]
```diff
@@ -252,9 +252,17 @@ describe('Validating rule docs', () => {
...
+          // Hard-code exceptions for code in headings.
+          if (
+            heading.text ===
+            'Explicit Resource Management (`await using` Statements)'
+          ) {
+            return;
+          }
+
+          expect(heading.text).toBe(titleCase(heading.text));
+        });
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/docs.test.ts:265]: Ehh, this is super specific to the heading. I think we'd be better off generalizing. How about removing anything inside code ticks with a regex?

```suggestion
        headings.forEach(heading => {
          const nonCodeText = heading.text.replace(/`[^`]*`/g, '');
          expect(nonCodeText).toBe(titleCase(nonCodeText));
        });
```

File: packages/eslint-plugin/tests/rules/await-thenable.test.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/await-thenable.test.ts:588]: No strong preference from me. I think it's fine as-is 🤷.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Clever, I like how the `getFixOrSuggest` change fits in!

Just requesting changes on the docs, WDYT?

Hugh Grant as Samuel Faulkner saying "oh, that's clever"

File: packages/eslint-plugin/tests/docs.test.ts (hunk 3 above)
Comment (kirkwaiblinger) [reply packages/eslint-plugin/tests/docs.test.ts:265]: oh, yeah, much better 👍

**Review by JoshuaKGoldberg — APPROVED** [review]
🙌

**Pull request merged**